keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1045 @@
+virtual TPM subsystem to provide virtual TPMs to pods
+
+**Be 100% compatible with software in the field today**: The primary example in this KEP is that the private key and certificate chain are delivered in a single credential-bundle file.  This is based on implementation and operation experience of a CSI-based certificates solution that delivered private keys and certificate chains in separate files, as expected by most commodity webservers today.  While this is an attractive idea, automatically rotating these separate files is fraught.  Even if both file's content is updated atomically, skewed reads in the workload can result in hard-to-diagnose errors if the workload reads the key before rotation and the certificate chain after rotation (or vice-versa).
+
+**Support private keys shared across multiple pods**: Sharing private keys and certificates across multiple pods may be required in certain applications.  However, this use case is well-served by Secret projected volume sources.
+
+**Support human-in-the-loop approval for certificate requests**: This KEP assumes that CSR approval and issuance is fully-automated, because this process blocks pod startup for pods that use PodCertificate projected volume sourcs.
+
+**Standardize on using the SPIFFE SVID format in core Kubernetes**: The certificates issued by `kubernetes.io/kube-apiserver-client-pod` can be made compatible with SPIFFE in the future, since they currently don't include any Subject Alternate Name entries.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.
+
+I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: pod-certificates-example
+spec:
+  restartPolicy: OnFailure
+  automountServiceAccountToken: false
+  containers:
+  - name: main
+    image: debian
+    command: ['sleep', 'infinity']
+    volumeMounts:
+    - name: kube-apiserver-client-certificate
+      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
+  volumes:
+  - name: kube-apiserver-client-certificate
+    projected:
+      sources:
+      - podCertificate:
+          signerName: "kubernetes.io/kube-apiserver-client-pod"
```

Comment:
> CN is username, Organizations are groups

This is where I thought we could use custom Subject RDN key names that I talked about during the api-review mtg. It might be a bit more flexible than an ASN.1 extension.

Reply:
OK, I'm about to upload a revision that adds a new PodCertificateRequest type.  I've left a bunch of UNRESOLVED blocks to update details of the signer.

Reply:
> This is where I thought we could use custom Subject RDN key names that I talked about during the api-review mtg - if such a need arose. It might be a bit more flexible than an ASN.1 extension.

I've done some googling, and I don't really understand what the usability difference in practice would be.  Both seem to require a custom OID, and non-standard parsing logic for the certificate.